asciidtype/asciidtype/src/casts.c

```diff
@@ -86,6 +87,190 @@ ascii_to_ascii_get_loop(PyArrayMethod_Context *context, int aligned,
     return 0;
 }
 
+static NPY_CASTING
+unicode_to_ascii_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                     PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                     PyArray_Descr *given_descrs[2],
+                                     PyArray_Descr *loop_descrs[2],
+                                     npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+    if (given_descrs[1] == NULL) {
+        Py_INCREF(given_descrs[0]);
+        loop_descrs[1] = given_descrs[0];
```

Comment:
This is not correct, I guess it may be a bit hard to reach right now.  `given_descrs[0]` will be a unicode dtype instance here, `loop_descr[1]` must be an ASCII dtype instance.

The correct thing will be to return an ASCII dtype instance.

N.B.: It would even be acceptable to always return an ASCII dtype instance of the same length as the unicode one (NumPy will chain the Unicode -> ASCII and ASCII -> ASCII cast, where the second one would change the size).

Reply:
Thank you for pointing this out, it was causing the issue I mentioned in the PR description, since `arr.astype(np.unicode_)` is going through `resolve_descriptors` to get the output dtype.